### ISSUE TITLE
Unable to download symbols on "old" versions

### ISSUE THREAD

**Issue by Rukkulakki** [microsoft/navcontainerhelper#1453]
After gathering courage for a while, I finally decided it was time to clean up all old images, and move all NAV/BC containers to the D drive. 
Following https://freddysblog.com/2020/10/10/bccontainerhelper-configuration/ to the letter, all images have been removed and I am now trying to recreate them on the d drive. Figuring I might as well start with the one that traditionally has caused problems, I picked the one that needs to run NAV2018. And sure enough, it is acting up.
Specifically, it won't let me download symbols. Running the client is fine, the old development client connects just fine. But VS Code is unable to connect to the development port. I've tried using the "generel" AL language extension, and installing the vsix from the 2018 Dev folder. Both fail

**Scripts used to create container and cause the issue**
```
$containerName = 'lederne'
$password = 'P@ssw0rd'
$securePassword = ConvertTo-SecureString -String $password -AsPlainText -Force
$credential = New-Object pscredential 'admin', $securePassword
$auth = 'UserPassword'
$artifactUrl = Get-BcArtifactUrl -type 'OnPrem' -version '11.0.24742.0' -country 'dk' -select 'Closest'
$licenseFile = 'c:\nav\infar\udvikling\nav2018.flf'
$ManagementServicesPort = 7050

$ClientServicesPort = $ManagementServicesPort + 1
$SOAPServicesPort = $ClientServicesPort + 1
$ODataServicesPort = $SOAPServicesPort + 1
$DeveloperServicesPort = $ODataServicesPort + 1
$WebPort = $DeveloperServicesPort + 1
$SSLPort = $WebPort + 1
$SQLPort = $SSLPort + 1

$additionalParameters = @(
        "-p published=$($ManagementServicesPort)-$DeveloperServicesPort,target=7045-7049",
        "-p published=$WebPort,target=80",
        "-p published=$SSLPort,target=443",
        "-p published=$SQLPort,target=1433"
    )

New-BcContainer `
    -accept_eula `
    -containerName $containerName `
    -credential $credential `
    -auth $auth `
    -artifactUrl $artifactUrl `
    -licenseFile $licenseFile `
    -includeCSIDE -doNotExportObjectsToText `
    -shortcuts CommonDesktopFolder `
    -installCertificateOnHost `
    -additionalParameters $additionalParameters `
    -updateHosts

Generate-SymbolsInNavContainer -containerName $containerName

Restart-BcContainer $containerName

**Full output of scripts**
BcContainerHelper is version 1.0.13-dev
BcContainerHelper is running as administrator
Host is Microsoft Windows 10 Enterprise - 20H2
Docker Client Version is 19.03.13
Docker Server Version is 19.03.13
Removing container lederne
Removing lederne from host hosts file
Removing lederne-* from host hosts file
Removing C:\ProgramData\BcContainerHelper\Extensions\lederne
Fetching all docker images
Using image mcr.microsoft.com/dynamicsnav:10.0.19042.630-generic
Creating Container lederne
Version: 11.0.24742.0-dk
Style: onprem
Multitenant: No
Platform: 11.0.24742.0
Generic Tag: 0.1.0.24
Container OS Version: 10.0.19042.630 (20H2)
Host OS Version: 10.0.19042.630 (20H2)
Using process isolation
Using locale da-DK
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file c:\nav\infar\udvikling\nav2018.flf
Additional Parameters:
-p published=7050-7054,target=7045-7049
-p published=7055,target=80
-p published=7056,target=443
-p published=7057,target=1433
Files in C:\ProgramData\BcContainerHelper\Extensions\lederne\my:
- AdditionalOutput.ps1
- AdditionalSetup.ps1
- license.flf
- MainLoop.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container lederne from image mcr.microsoft.com/dynamicsnav:10.0.19042.630-generic
6d5a60c10fbe68ccdac35df6d855fd8590b2b1d24cb676c848f4cc813dd24995
Waiting for container lederne to be ready
Using artifactUrl https://bcartifacts.azureedge.net/onprem/11.0.24742.0/dk
Using installer from C:\Run\110
Copying Platform Artifacts
Copying Application Artifacts
Copying Database
Copying Installers
Copying ConfigurationPackages
Copying TestToolKit
Copying UpgradeToolKit
Copying Extensions
Installing NAV
Installing Url Rewrite
Installing OpenXML
Installing DotNetCore
Starting Local SQL Server
Starting Internet Information Server
Copying Service Tier Files
Copying Web Client Files
Copying Windows Client Files
Copying PowerShell Scripts
Copying ConfigurationPackages
Copying Test Assemblies
Copying TestToolKit
Copying UpgradeToolKit
Copying Extensions
Copying ClientUserSettings
Copying ReportBuilder
Restoring CRONUS Demo Database
Determining Database Collation
Changing Database Server Collation to Danish_Greenlandic_100_CS_AS
Restoring CRONUS Demo Database
Installing C:\NAVDVD\Installers\DK\RTC\RoleTailoredClient.Local.Dk.msi
Skipping C:\NAVDVD\Installers\DK\WebHelp\WebHelp.Local.Dk.msi
Modifying NAV Service Tier Config File for Docker
Creating NAV Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 111 seconds
Installation complete
Initializing...
Setting host.docker.internal to 10.4.28.52 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 10.4.28.52 in container hosts file (copy from host hosts file)
Setting kubernetes.docker.internal to 127.0.0.1 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 172.23.224.1 in container hosts file
Starting Container
Hostname is lederne
PublicDnsName is lederne
Using NavUserPassword Authentication
Creating Self Signed Certificate
Self Signed Certificate Thumbprint 30CC8237387EB79BEE7B28ABF0A35D85630FAE8D
Modifying Service Tier Config File with Instance Specific Settings
Starting Service Tier
Registering event sources
Creating DotNetCore Web Server Instance
Using license file 'c:\run\my\license.flf'
Import License
Creating http download site
Setting SA Password and enabling SA
Creating admin as SQL User and add to sysadmin
Creating SUPER user
Container IP Address: 172.23.227.49
Container Hostname  : lederne
Container Dns Name  : lederne
Web Client          : http://lederne/NAV/
Dev. Server         : http://lederne
Dev. ServerInstance : NAV
Setting lederne to 172.23.227.49 in host hosts file

Files:
http://lederne:8080/ALLanguage.vsix

Container Total Physical Memory is 63.9Gb
Container Free Physical Memory is 45.3Gb

Initialization took 56 seconds
Ready for connections!
Reading CustomSettings.config from lederne
Creating Desktop Shortcuts for lederne
Container lederne successfully created

Use:
Get-BcContainerEventLog -containerName lederne to retrieve a snapshot of the event log from the container
Get-BcContainerDebugInfo -containerName lederne to get debug information about the container
Enter-BcContainer -containerName lederne to open a PowerShell prompt inside the container
Remove-BcContainer -containerName lederne to remove the container again
docker logs lederne to retrieve information about URL's again
Creating application symbols in lederne
[0] [13-11-20 11:25:06] The command completed successfully in '126' seconds.
Symbols successfully generated
lederne
Waiting for container lederne to be ready

Initializing...
Setting host.docker.internal to 10.4.28.52 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 10.4.28.52 in container hosts file (copy from host hosts file)
Setting kubernetes.docker.internal to 127.0.0.1 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 172.23.224.1 in container hosts file
Restarting Container
PublicDnsName unchanged
Hostname is lederne
PublicDnsName is lederne
Using NavUserPassword Authentication
Starting Local SQL Server
Starting Internet Information Server
Starting Service Tier
Container IP Address: 172.23.228.19
Container Hostname  : lederne
Container Dns Name  : lederne
Web Client          : http://lederne/NAV/
Dev. Server         : http://lederne
Dev. ServerInstance : NAV
Setting lederne to 172.23.228.19 in host hosts file

Files:
http://lederne:8080/ALLanguage.vsix

Container Total Physical Memory is 63.9Gb
Container Free Physical Memory is 46.5Gb

Initialization took 3 seconds
Ready for connections!

** From VS Code **
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "al",
            "request": "launch",
            "name": "Rukkulakki",
            "server": "http://lederne",
            "serverInstance": "NAV",
            "port": 7054,
            "authentication": "UserPassword",
            "startupObjectId": 22
        }
    ]
}

[2020-11-13 12:34:07.74] Using reference symbols cache path: d:\Git\Lederne\Time Sheet\.alpackages
[2020-11-13 12:34:07.78] Sending request to http://lederne:7054/NAV/dev/packages?publisher=Microsoft&appName=Application&versionText=11.0.0.0
[2020-11-13 12:34:07.79] Sending request to http://lederne:7054/NAV/dev/packages?publisher=Microsoft&appName=System&versionText=11.0.0.0
[2020-11-13 12:34:09.81] Error: Could not connect to the server. Please verify that:
1. The server is started.
2. The development endpoint is enabled (DeveloperServicesEnabled server setting).
3. The service name matches the one specified in the URL.
4. The port number matches the one specified in the URL (DeveloperServicesPort server setting).
5. The protocol matches the one specified in the URL (DeveloperServicesSSLEnabled server setting), and that a trusted certificate is used in case of an SSL connection.
[2020-11-13 12:34:09.81] Error: Could not connect to the server. Please verify that:
1. The server is started.
2. The development endpoint is enabled (DeveloperServicesEnabled server setting).
3. The service name matches the one specified in the URL.
4. The port number matches the one specified in the URL (DeveloperServicesPort server setting).
5. The protocol matches the one specified in the URL (DeveloperServicesSSLEnabled server setting), and that a trusted certificate is used in case of an SSL connection.

**Screenshots**
![Lederne - Visual Studio Code](https://user-images.githubusercontent.com/25152707/99068006-fbdafc00-25ab-11eb-86c0-fa5963b4c3f9.png)

**Comment by freddydk** [comment]
What is the reason for the port mappings?
We you going to access the container from other machines?

**Comment by Rukkulakki** [comment]
Just easier as I might have several containers running, and I don't want them communicating on the same port.

**Comment by freddydk** [comment]
Could you try to remove the port mappings and test?

**Comment by Rukkulakki** [comment]
Remove the entire additionsettings?

**Comment by freddydk** [comment]
yes - the -additionalParameters line (remember you cannot outcomment one line in a multi-line PS command - you need to remove it)

**Comment by Rukkulakki** [comment]
It works without the mapping!

But the old container that was on the C drive was also mapped, and that worked. Admittedly it has been a while since I touched it, sp it can just as easily be something caused by an windows/docker update

**Comment by freddydk** [comment]
Hmmm 

the -p parameter maps a port from the container on the host - meaning that you can connect through the host to the new port.
it doesn't change the port on the container - never has.
Maybe you have used localhost and the new port.

Recommended way is to use different container names in vs code - not ports.
If you need to use ports - recommended way is to change the port in the container - not use mapping (-DeveloperServicesPort 7066)
If you need multiple containers on the same machine (published) the recommended solution is traefik.
But you can do what you did - you just need to connect to the host instead of the container.

This blog post from 2018 explains this a bit https://freddysblog.com/2018/09/03/navcontainerhelper-0-3-1-0-and-a-new-docker-generic-build-0-0-6-6/

**Comment by Rukkulakki** [comment]
Maybe. I only checked Docker, and can't remember how the launch file started.

Looks like the -p parameter has been added to the script somewhere along the way. Looking at some of the older scripts, it is not present. it was just "--publish 7046-7049:7046-7049" to open the ports wide. Not sure why it has been changed ...

Anyhow: I just tried it with just using the -DeveloperServicePort  and it works great in VS Code.

So now the script reads:
$containerName = 'lederne'
$password = 'P@ssw0rd'
$securePassword = ConvertTo-SecureString -String $password -AsPlainText -Force
$credential = New-Object pscredential 'admin', $securePassword
$auth = 'UserPassword'
$artifactUrl = Get-BcArtifactUrl -type 'OnPrem' -version '11.0.24742.0' -country 'dk' -select 'Closest'
$licenseFile = 'c:\nav\infar\udvikling\nav2018.flf'
[int]$ManagementServicesPort = 7050

$ClientServicesPort = $ManagementServicesPort + 1
$SOAPServicesPort = $ClientServicesPort + 1
$ODataServicesPort = $SOAPServicesPort + 1
$DeveloperServicesPort = $ODataServicesPort + 1

New-BcContainer `
    -accept_eula `
    -containerName $containerName `
    -credential $credential `
    -auth $auth `
    -artifactUrl $artifactUrl `
    -licenseFile $licenseFile `
    -includeCSIDE -doNotExportObjectsToText `
    -shortcuts CommonDesktopFolder `
    -installCertificateOnHost `
    -WebClientPort $ClientServicesPort `
    -ManagementServicesPort $ManagementServicesPort `
    -SoapServicesPort $SOAPServicesPort `
    -ODataServicesPort $ODataServicesPort `
    -DeveloperServicesPort $DeveloperServicesPort `
    -updateHosts

Generate-SymbolsInNavContainer -containerName $containerName

Restart-BcContainer $containerName

And I can connect to it from client, old dev, and VS Code 👍 
Haven't tried connecting from a seperate machine, but for now I am happy, and the issue can be closed.
